### PR TITLE
Feature 10 - Sound

### PR DESCRIPTION
**Pull request by AihasasYadu**
Added Audio for:- 
>Menu Button Clicks
>Character's Landing
>BG Music for Lobby and Levels
>Game Over Overlay
>Level Complete Overlay
>Key Collection

### CODE REVIEW

File: Assets/Game Assets/AY_Scripts/Audio/AudioManagerController.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            s.source.pitch = s.pitch;
+            s.source.loop = s.loop;
+
+        }   
+    }
+    private void Start()
+    {
+        Play(themeMusicName);
+    }
+    public void Play(string name)
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/Audio/AudioManagerController.cs]: enum

File: Assets/Game Assets/AY_Scripts/Audio/AudioManagerController.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        if (instance != null)
+        {
+            Destroy(gameObject);
+        }
+        else
+        {
+            instance = this;
+            DontDestroyOnLoad(gameObject);
+        }
+        foreach (Sound s in sounds)
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/Audio/AudioManagerController.cs]: few sounds sources - sources can be static (setup one time)

File: Assets/Game Assets/AY_Scripts/Audio/AudioManagerController.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    public void Play(string name)
+    {
+        Sound s = Array.Find(sounds, sound => sound.name == name);
+        if(s == null)
+        {
+            Debug.Log("Incorrect File Name/File Not Found : " + name);
+            return;
+        }
+        s.source.Play();
+    }
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/Audio/AudioManagerController.cs:44]: void Stop(Sound sound) { 
.... 
}

Comment (mayankgrover) [reply Assets/Game Assets/AY_Scripts/Audio/AudioManagerController.cs:44]: void StopAll* 
StopAllEffects()
StopTheme(){ ... }

File: Assets/Game Assets/AY_Scripts/GameOverController.cs [hunk 4]
```diff
@@ -38,9 +37,12 @@ public void GameOverOverlay()
...
         if (!gameOverIMG_Positoned)
         {
-            gameOverAnim.Play("GameOverEntrance", -1, 0f);
-            Debug.Log("Positioned");
+            gameOverAnim.Play(gameOverEntryAnim, -1, 0f);
             gameOverIMG_Positoned = !gameOverIMG_Positoned;
+            AudioManagerController a = FindObjectOfType<AudioManagerController>();
+            a.Play(gameOverSound);
+            a.sounds[0].source.volume = 0;
+            Destroy(AudioManagerController.Instance);
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/GameOverController.cs]: get rid of this

File: Assets/Game Assets/AY_Scripts/HomeButtonController.cs [hunk 5]
```diff
@@ -7,23 +7,14 @@
...
-    {
-        if (buttonClicked)
-        {
-            LoadLobby();
-        }
+        button.onClick.AddListener(LoadLobby);
     }
     private void LoadLobby()
     {
+        Destroy(AudioManagerController.Instance);
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/HomeButtonController.cs]: remove

File: Assets/Game Assets/AY_Scripts/Level_Over_Controller.cs [hunk 6]
```diff
@@ -7,11 +7,16 @@
 public class Level_Over_Controller : MonoBehaviour
 {
     public Image levelFinished;
+    public Image levelCompleteGIF;
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/Level_Over_Controller.cs]: fix

File: Assets/Game Assets/AY_Scripts/Level_Over_Controller.cs [hunk 7]
```diff
@@ -38,6 +43,11 @@ private void LevelCompleteOverlayTransition()
         {
             pos.x = 0;
             overlayPositioned = true;
+            levelCompleteGIF.GetComponent<Animator>().SetBool(levelCompleteAnimParameter, true);
+            AudioManagerController a = FindObjectOfType<AudioManagerController>();
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/Level_Over_Controller.cs]: fix

File: Assets/Game Assets/AY_Scripts/Levels/LevelLoader.cs [hunk 8]
```diff
@@ -33,8 +40,10 @@ private void LoadCheck()
 
     private void LoadLevel()
     {
-        PlayerPrefs.SetInt("PlayerHealth", 100);
-        PlayerPrefs.SetInt("CurrentScore", 0);
+        PlayerPrefs.SetInt(prefabHealth, 100);
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/Levels/LevelLoader.cs:42]: hardcoding 
figure out better place for the health update logic

File: Assets/Game Assets/AY_Scripts/Levels/LevelLoader.cs [hunk 9]
```diff
@@ -33,8 +40,10 @@ private void LoadCheck()
...
+        PlayerPrefs.SetInt(prefabScore, 0);
+        FindObjectOfType<AudioManagerController>().sounds[0].source.volume = 0;
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/Levels/LevelLoader.cs]: AudioManagerController.Instance.SetVolume(AudioType someref, float volume)

File: Assets/Game Assets/AY_Scripts/LobbyButtonsController.cs [hunk 10]
```diff
@@ -10,28 +10,31 @@ public class LobbyButtonsController : MonoBehaviour, IPointerEnterHandler, IPoin
...
+        button.onClick.AddListener(OutOfCanvas);
     }
     private void Update()
     {
         if (!outOfCanvas)
         {
             Move();
         }
-        if(button.transform.position.x > 1000)
+        if(button.transform.position.x > 2000)
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/LobbyButtonsController.cs]: itween

File: Assets/Game Assets/AY_Scripts/LobbyButtonsController.cs [hunk 11]
```diff
@@ -40,11 +43,11 @@ public void Move()
     }
     public void OnPointerEnter(PointerEventData data)
     {
-        GetComponent<Animator>().SetBool("MouseHover", true);
+        GetComponent<Animator>().SetBool(mouseHoverAnimCondition, true);
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/LobbyButtonsController.cs]: cache

File: Assets/Game Assets/AY_Scripts/LobbyController.cs [hunk 12]
```diff
@@ -19,21 +23,30 @@ private void Awake()
...
     {
-        int level = 1;
-        PlayerPrefs.SetInt("PlayerHealth", 100);
-        PlayerPrefs.SetInt("CurrentScore", 0);
-        level = PlayerPrefs.GetInt("LastLevel");
+        AudioManagerController a = FindObjectOfType<AudioManagerController>();
+        a.Play(menuButtonClickSound);
+        a.sounds[0].source.volume = 0;
+        Destroy(AudioManagerController.Instance);
+        PlayerPrefs.SetInt(prefabHealth, 100);
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/LobbyController.cs]: dup

File: Assets/Game Assets/AY_Scripts/PlayerController.cs [hunk 13]
```diff
@@ -188,25 +205,28 @@ public void DecreaseHealth(int decreaseHealthBy)
 
     private void KillPlayer()
     {
-        anime.Play("Ellen_Death", -1, 0f);
-        anime.SetBool("Death", true);
+        anime.Play(deathAnimName, -1, 0f);
+        anime.SetBool(deathAnimParameter, true);
         FreezePlayer();
-        PlayerPrefs.SetInt("LastLevel", SceneManager.GetActiveScene().buildIndex);
+        PlayerPrefs.SetInt(prefabLevel, SceneManager.GetActiveScene().buildIndex);
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/PlayerController.cs]: SaveController

File: Assets/Game Assets/AY_Scripts/RestartButtonController.cs [hunk 14]
```diff
@@ -8,26 +8,24 @@
...
-        PlayerPrefs.SetInt("CurrentScore", 0);
+        PlayerPrefs.SetInt(prefabHealth, 100);
+        PlayerPrefs.SetInt(prefabScore, 0);
         Scene currentScene = SceneManager.GetActiveScene();
+        SetThemeAudioVolume();
         SceneManager.LoadScene(currentScene.buildIndex);
     }
+    private void SetThemeAudioVolume()
+    {
+        if(AudioManagerController.Instance != null)
```
Comment (mayankgrover) [inline Assets/Game Assets/AY_Scripts/RestartButtonController.cs]: not needed

**Review by mayankgrover — CHANGES_REQUESTED** [review]

**Comment by AihasasYadu** [comment]
replaced manual animation with iTween

**Pull request merged**